modules/fedimint-mint-client/src/repair_wallet.rs

```diff
@@ -0,0 +1,146 @@
+                    let spent = retry("fetch e-cash spentness", aggressive_backoff(), || async {
+                        Ok(module_api_inner.check_note_spent(key.nonce).await?)
+                    })
+                    .await?;
+                    anyhow::Ok(if spent { Some(key) } else { None })
+                }
+            })
+            .buffer_unordered(CHECK_PARALLELISM)
+            .try_filter_map(|result| async move { Ok(result) })
+            .try_collect()
+            .await?;
+
+        for note_key in spent_notes {
+            summary.spent_notes.inc(note_key.amount, 1);
+            dbtx.remove_entry(&note_key).await;
+        }
+
+        let next_indices: BTreeMap<_, _> = {
+            let mut db_next_indexes = dbtx
+                .find_by_prefix_sorted_descending(&NextECashNoteIndexKeyPrefix)
+                .await
+                .map(|(key, idx)| (key.0, idx))
+                .collect::<BTreeMap<_, _>>()
+                .await;
+
+            self.cfg
+                .tbs_pks
+                .tiers()
+                .map(|&denomination| {
+                    (
+                        denomination,
+                        db_next_indexes.remove(&denomination).unwrap_or_default(),
+                    )
+                })
+                .collect()
+        };
+
+        // Next check if any of the indices for issuing new notes are already used
+        let used_nonces = stream::iter(next_indices.into_iter())
+            .map(|(amount, original_next_index)| {
+                let module_api_inner = module_api.clone();
+                async move {
+                    let mut next_index = original_next_index;
+                    let maybe_advanced_index = loop {
+                        let note_secret = Self::new_note_secret_static(
+                            &self.secret,
+                            amount,
+                            NoteIndex(next_index),
+                        );
+                        let (_, blind_nonce) = NoteIssuanceRequest::new(&self.secp, &note_secret);
+                        let nonce_used = retry(
+                            "checking if blind nonce was already used",
+                            aggressive_backoff(),
+                            || async {
+                                Ok(module_api_inner.check_blind_nonce_used(blind_nonce).await?)
+                            },
+                        )
+                        .await?;
+
+                        if nonce_used {
+                            // If the nonce was already used, try again with the next index
+                            next_index += 1;
+                        } else if original_next_index == next_index {
+                            // If the initial nonce wasn't used we are good, nothing to be done
+                            break None;
+                        } else {
+                            // If the initial nonce was used but we found an unused one by now,
```

Comment:
I don't think there's any guarantee that every nonce that was allocated will be used, and that's why in recovery we need to use a gap. I guess it's OK for a checking tool to check only one ahead, especially due to privacy, but worth keeping in mind.